schemes/hack_shallow/cmfmca.F90

```diff
@@ -0,0 +1,978 @@
+  real(kind_phys) :: ssfac = 1.001_kind_phys    ! detrained air supersaturation bound [???]
+
+  ! internal parameters for tolerance
+  real(kind_phys) :: tiny = 1.0e-36_kind_phys   ! arbitrary small num in scalar transport estimates
+  real(kind_phys) :: eps  = 1.0e-13_kind_phys   ! machine dependent convergence criteria
+  real(kind_phys) :: tpmax = 1.50_kind_phys     ! maximum acceptable T perturbation [K]
+  real(kind_phys) :: shpmax = 1.50e-3_kind_phys ! maximum acceptable Q perturbation [g/g]
+
+  ! diagnostic only
+  logical         :: debug_verbose = .false.    ! control for debug messages
+
+
+contains
+  ! Initialization of moist convective mass procedure including namelist read.
+!> \section arg_table_cmfmca_init Argument Table
+!! \htmlinclude cmfmca_init.html
+  subroutine cmfmca_init( &
+    pver, &
+    cmftau_in, c0_in, &
+    rair, cpair, gravit, latvap, rhoh2o_in, &
+    pref_edge, &
+    errmsg, errflg &
+  )
+
+    integer,            intent(in)  :: pver         ! number of vertical levels
+    real(kind_phys),    intent(in)  :: cmftau_in    ! characteristic adjustment time scale [s]
+    real(kind_phys),    intent(in)  :: c0_in        ! rain water autoconversion coefficient [m-1]
+    real(kind_phys),    intent(in)  :: rair         ! gas constant for dry air [J K-1 kg-1]
+    real(kind_phys),    intent(in)  :: cpair        ! specific heat of dry air [J K-1 kg-1]
+    real(kind_phys),    intent(in)  :: gravit       ! gravitational constant [m s-2]
+    real(kind_phys),    intent(in)  :: latvap       ! latent heat of vaporization [J kg-1]
+    real(kind_phys),    intent(in)  :: rhoh2o_in    ! density of liquid water [kg m-3]
+    real(kind_phys),    intent(in)  :: pref_edge(:) ! reference pressures at interface [Pa]
+
+    character(len=512), intent(out) :: errmsg
+    integer,            intent(out) :: errflg
+
+    ! local variables
+    integer :: k
+
+    errmsg = ''
+    errflg = 0
+
+    ! namelist variables
+    cmftau  = cmftau_in
+    c0      = c0_in
+
+    ! host model physical constants
+    cp      = cpair
+    rcp     = 1.0_kind_phys/cp
+    hlat    = latvap
+    rhlat   = 1.0_kind_phys/hlat
+    grav    = gravit
+    rgrav   = 1.0_kind_phys/gravit
+    rgas    = rair
+    rhoh2o  = rhoh2o_in
+
+    ! determine limit of shallow convection: regions below 40 mb
+    ! logic ported from convect_shallow_init with note that this calculation is repeated in the deep
+    ! convection interface.
+    if(pref_edge(1) >= 4.e3_kind_phys) then
+      limcnv = 1
+    else
+      limcnv = pver + 1
+      do k = 1, pver
+        if(pref_edge(k) < 4.e3_kind_phys .and. pref_edge(k+1) >= 4.e3_kind_phys) then
+          limcnv = k
+        endif
+      enddo
+    endif
+
+    ! in-module parameters (hardcoded)
+    dzmin = 0.0_kind_phys
+    betamn = 0.10_kind_phys
+
+    ! specify that relaxation timescale should be applied to column as opposed to triplets individually
+    rlxclm = .true.
```

Comment:
can you just set these above at the module level, as you do `ssfac`, etc?

Reply:
Thanks, done!